# HG changeset patch
# User alta88@fixall.com
# Date 1746552134 21600
# Parent  d32270f2ba2d13d8ac1b55fe77f57c75ad0a586a
Feature: Add Horizontal layout view to 3pane.
- Bug 1964119

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -6777,8 +6777,11 @@ commandController.registerCallback("cmd_
 );
 commandController.registerCallback("cmd_viewStackedMailLayout", () =>
   Services.prefs.setIntPref("mail.pane_config.dynamic", 4)
 );
+commandController.registerCallback("cmd_viewHorizontalMailLayout", () =>
+  Services.prefs.setIntPref("mail.pane_config.dynamic", 5)
+);
 commandController.registerCallback(
   "cmd_toggleThreadPaneHeader",
   () => threadPaneHeader.toggleThreadPaneHeader(),
   () => gFolder && !gFolder.isServer
diff --git a/mail/base/content/mainCommandSet.inc.xhtml b/mail/base/content/mainCommandSet.inc.xhtml
--- a/mail/base/content/mainCommandSet.inc.xhtml
+++ b/mail/base/content/mainCommandSet.inc.xhtml
@@ -48,8 +48,9 @@
      <command id="cmd_viewWideMailLayout" oncommand="goDoCommand('cmd_viewWideMailLayout')" disabled="true"/>
      <command id="cmd_viewVerticalMailLayout" oncommand="goDoCommand('cmd_viewVerticalMailLayout')" disabled="true"/>
      <command id="cmd_viewWideThreadMailLayout" oncommand="goDoCommand('cmd_viewWideThreadMailLayout')" disabled="true"/>
      <command id="cmd_viewStackedMailLayout" oncommand="goDoCommand('cmd_viewStackedMailLayout')" disabled="true"/>
+     <command id="cmd_viewHorizontalMailLayout" oncommand="goDoCommand('cmd_viewHorizontalMailLayout')" disabled="true"/>
      <command id="cmd_toggleFolderPane" oncommand="goDoCommand('cmd_toggleFolderPane')" disabled="true"/>
      <command id="cmd_toggleThreadPaneHeader" oncommand="goDoCommand('cmd_toggleThreadPaneHeader')" disabled="true"/>
      <command id="cmd_toggleMessagePane" oncommand="goDoCommand('cmd_toggleMessagePane')" disabled="true"/>
      <command id="cmd_viewAllMsgs" oncommand="goDoCommand('cmd_viewAllMsgs')" disabled="true"/>
diff --git a/mail/base/content/messenger-menubar.inc.xhtml b/mail/base/content/messenger-menubar.inc.xhtml
--- a/mail/base/content/messenger-menubar.inc.xhtml
+++ b/mail/base/content/messenger-menubar.inc.xhtml
@@ -366,8 +366,13 @@
                   type="radio"
                   data-l10n-id="menu-view-layout-stacked"
                   name="viewlayoutgroup"
                   command="cmd_viewStackedMailLayout"/>
+        <menuitem id="messagePaneHorizontal"
+                  type="radio"
+                  data-l10n-id="menu-view-layout-horizontal"
+                  name="viewlayoutgroup"
+                  command="cmd_viewHorizontalMailLayout"/>
         <menuseparator id="viewMenuAfterPaneVerticalSeparator"/>
         <menuitem id="menu_showFolderPane" type="checkbox" label="&showFolderPaneCmd.label;"
                   accesskey="&showFolderPaneCmd.accesskey;" command="cmd_toggleFolderPane"/>
         <menuitem id="menu_toggleThreadPaneHeader"
diff --git a/mail/base/content/widgets/pane-layout.mjs b/mail/base/content/widgets/pane-layout.mjs
--- a/mail/base/content/widgets/pane-layout.mjs
+++ b/mail/base/content/widgets/pane-layout.mjs
@@ -40,9 +40,9 @@ class PaneLayout extends HTMLBodyElement
    * Maps the preference integer into a layout name array.
    *
    * @type {Array}
    */
-  LAYOUTS = ["standard", "wide", "vertical", "widethread", "stacked"];
+  LAYOUTS = ["standard", "wide", "vertical", "widethread", "stacked", "horizontal"];
 
   /**
    * Maps the layout name properties for dom node attributes.
    *
@@ -73,8 +73,13 @@ class PaneLayout extends HTMLBodyElement
       className: "layout-stacked",
       folderPaneSplitterDir: "vertical",
       messagePaneSplitterDir: "horizontal",
     },
+    horizontal: {
+      className: "layout-horizontal",
+      folderPaneSplitterDir: "vertical",
+      messagePaneSplitterDir: "vertical",
+    },
   };
 
   /**
    * If the account central browser is currently visible.
diff --git a/mail/components/customizableui/content/panelUI.inc.xhtml b/mail/components/customizableui/content/panelUI.inc.xhtml
--- a/mail/components/customizableui/content/panelUI.inc.xhtml
+++ b/mail/components/customizableui/content/panelUI.inc.xhtml
@@ -359,8 +359,14 @@
                        type="radio"
                        data-l10n-id="menu-view-layout-stacked"
                        name="viewlayoutgroup"
                        command="cmd_viewStackedMailLayout"/>
+        <toolbarbutton id="appmenu_messagePaneHorizontal"
+                       class="subviewbutton subviewbutton-iconic"
+                       type="radio"
+                       data-l10n-id="menu-view-layout-horizontal"
+                       name="viewlayoutgroup"
+                       command="cmd_viewHorizontalMailLayout"/>
         <toolbarseparator id="appmenu_viewMenuAfterPaneVerticalSeparator"/>
         <toolbarbutton id="appmenu_showFolderPane"
                        class="subviewbutton subviewbutton-iconic"
                        type="checkbox"
diff --git a/mail/components/extensions/parent/ext-mailTabs.js b/mail/components/extensions/parent/ext-mailTabs.js
--- a/mail/components/extensions/parent/ext-mailTabs.js
+++ b/mail/components/extensions/parent/ext-mailTabs.js
@@ -29,9 +29,17 @@ XPCOMUtils.defineLazyPreferenceGetter(
   "mail.pane_config.dynamic",
   0
 );
 
-const LAYOUTS = ["standard", "wide", "vertical", "widethread", "stacked"];
+const LAYOUTS = [
+  "standard",
+  "wide",
+  "vertical",
+  "widethread",
+  "stacked",
+  "horizontal",
+];
+
 // From nsIMsgDBView.idl
 const SORT_TYPE_MAP = new Map(
   Object.keys(Ci.nsMsgViewSortType).map(key => {
     // Change "byFoo" to "foo".
diff --git a/mail/components/extensions/schemas/mailTabs.json b/mail/components/extensions/schemas/mailTabs.json
--- a/mail/components/extensions/schemas/mailTabs.json
+++ b/mail/components/extensions/schemas/mailTabs.json
@@ -163,9 +163,9 @@
           "layout": {
             "type": "string",
             "description": "Sets the arrangement of the folder pane, message list pane, and message display pane. Note that setting this applies it to all mail tabs.",
             "optional": true,
-            "enum": ["standard", "wide", "vertical", "widethread", "stacked"]
+            "enum": ["standard", "wide", "vertical", "widethread", "stacked", "horizontal"]
           },
           "folderPaneVisible": {
             "type": "boolean",
             "description": "Shows or hides the folder pane.",
diff --git a/mail/locales/en-US/messenger/menubar.ftl b/mail/locales/en-US/messenger/menubar.ftl
--- a/mail/locales/en-US/messenger/menubar.ftl
+++ b/mail/locales/en-US/messenger/menubar.ftl
@@ -132,8 +132,12 @@ menu-view-layout-widethread =
 menu-view-layout-stacked =
     .label = Stacked View
     .accesskey = S
 
+menu-view-layout-horizontal =
+    .label = Horizontal View
+    .accesskey = z
+
 menu-view-toggle-thread-pane-header =
     .label = Message List Header
     .accesskey = H
 
diff --git a/mail/themes/shared/mail/about3Pane.css b/mail/themes/shared/mail/about3Pane.css
--- a/mail/themes/shared/mail/about3Pane.css
+++ b/mail/themes/shared/mail/about3Pane.css
@@ -117,8 +117,17 @@ body {
                    "threads messagePaneSplitter message" minmax(auto, 1fr)
                    / minmax(auto, 1fr) min-content minmax( var(--messagePaneSplitter-width), 1fr);
   }
 
+  /* Horizontal layout: folder tree and thread tree and message browser one after the other. */
+  &.layout-horizontal {
+    grid-template: "folders" var(--folderPaneSplitter-height)
+    "folderPaneSplitter" min-content
+    "threads" auto
+    "messagePaneSplitter" min-content
+    "message" var(--messagePaneSplitter-height);
+  }
+
   /* If Account Central is shown, it overrides the layout setting. */
   &.account-central {
     grid-template: "folders folderPaneSplitter account-central" auto
                    / minmax(auto, var(--folderPaneSplitter-width)) min-content minmax(auto, 1fr);
